feat: implement exercise #1

NguyenTienDat62th1 opened the pull request:
- Khởi tạo project
- Chức năng: Đăng nhập, Đăng ký, Quên mật khẩu

NguyenTienDat62th1 committed:
feat: implement exercise #1

hungnm-nws commented on src/app.module.ts:
```diff
@@ -0,0 +1,18 @@
+  imports: [MongooseModule.forRoot(mongoConfig.uri), AuthModule, JwtModule.register({
+    global: true,
+    secret: 'secretKey',
+    signOptions: { expiresIn: '1h' },
```
các config ở đây, uri, secret, expiresIn cho ra hết env

hungnm-nws commented on src/auth/auth.controller.ts:
```diff
@@ -0,0 +1,58 @@
+    },
+  })
+  @ApiResponse({ status: 201, description: 'User registered successfully' })
+  async register(@Body() body: any) {
```
không dùng any

hungnm-nws commented on src/auth/auth.controller.ts:
```diff
@@ -0,0 +1,58 @@
+    // Tạo token reset password (ở đây demo đơn giản, bạn có thể generate JWT hoặc random string)
+    const resetToken = Math.random().toString(36).substring(2);
+
+    await this.authService.sendResetPasswordMail(body.email, resetToken);
```
bê xử lý vào tầng service

hungnm-nws commented on src/auth/auth.controller.ts:
```diff
@@ -0,0 +1,58 @@
+      };
+    }
+    // Tạo token reset password (ở đây demo đơn giản, bạn có thể generate JWT hoặc random string)
+    const resetToken = Math.random().toString(36).substring(2);
```
token này để làm gì ? tìm hiểu lại cách xử lý, mục đích của token

hungnm-nws commented on src/auth/auth.module.ts:
```diff
@@ -0,0 +1,20 @@
+    JwtModule.register({
+    global: true,
+    secret: 'secretKey',
+    signOptions: { expiresIn: '1h' },
```
bê config ra env

hungnm-nws commented on src/auth/auth.service.ts:
```diff
@@ -0,0 +1,77 @@
+
+  async validatePassword(password: string, hash: string) {
+    return await bcrypt.compare(password, hash);
+  }
```
viết 1 file helper dùng chung để chứa các hàm này

hungnm-nws commented on src/auth/auth.service.ts:
```diff
@@ -0,0 +1,77 @@
+      to: email,
+      subject: 'Reset Password',
+      text: `Click here to reset password: http://localhost:3000/reset?token=${token}`,
+    });
```
tách ra mail service riêng, auth service chỉ xử lý nghiệp vụ liên quan đến auth

hungnm-nws commented on src/auth/auth.service.ts:
```diff
@@ -0,0 +1,77 @@
+      return {
+        message: 'Email is required',
+      };
+    }
```
phải check mail có trong hệ thống ko thì mới gửi chứ

hungnm-nws commented on src/config/mongo.config.ts:
```diff
@@ -0,0 +1,5 @@
+import { MongooseModuleOptions } from '@nestjs/mongoose';
+
+export const mongoConfig = {
+  uri: 'mongodb://localhost:27017/nestdb',
```
dùng env

hungnm-nws commented on src/user/schemas/user.schema.ts:
```diff
@@ -0,0 +1,13 @@
+import { Prop, Schema, SchemaFactory } from '@nestjs/mongoose';
+import { Document } from 'mongoose';
+
+@Schema()
```
thêm timestamp, thêm 1 số trường cho phong phú thêm

hungnm-nws commented on src/user/schemas/user.schema.ts:
```diff
@@ -0,0 +1,13 @@
+@Schema()
+export class User extends Document {
+  @Prop({ unique: true })
+  username: string;
```
hệ thống cần email để làm các hàm gửi mail, thêm hoặc đổi tên username

hungnm-nws commented on src/user/user.controller.ts:
```diff
@@ -0,0 +1,33 @@
+    @Get('profile')
+    @UseGuards(JwtAuthGuard)
+    @ApiBearerAuth()
+    getProfile(@CurrentUser() user: any) {
```
không dùng any

NguyenTienDat62th1 committed:
feat(auth): refactor forgot password flow
- move config (uri, secret, expiresIn) to .env
- remove any, apply DTOs
- shift handling logic to service
- add timestamp and extra fields
- adjust forgot password logic

NguyenTienDat62th1 committed:
fix exercise1

NguyenTienDat62th1 committed:
fix: add api reset password

NguyenTienDat62th1 committed:
fix: token
